test/test_evaluator.cpp

```diff
@@ -417,4 +417,22 @@ TEST_F(ExprEvalTest, TensorMetaData) {
   checkIntValue(evaluator, stride1, 1L);
 }
 
+TEST_F(ExprEvalTest, OpaqueEquality) {
+  Opaque a{DataType::Int}, b{DataType::Int};
+  // Because the type information is not available at compile time, we can't
+  // accurately compare the values of two opaque values. So, by default,
+  // equality check is done by pointer compare.
+  EXPECT_EQ(a, a);
+  EXPECT_EQ(b, b);
+  EXPECT_NE(a, b);
+  EXPECT_NE(b, a);
+  // However, we can manually specify the comparator to use for equality check.
+  Opaque c{DataType::Int, OpaqueEquals<PrimDataType>{}},
```

Comment:
I think since `a` and `c` have different `.equals` this makes equality assymetric: `a == c` is `false` but `c == a` is `true`. In order to make this symmetric, should `operator==()` instead use `this->equals(*this, other) || other.equals(*this, other)`?

Reply:
That's an interesting question. Also, what happens if two Opaque values have different equality methods? Are they considered comparable?

Reply:
I think if two opaque values have different equality methods, this should be considered as an error. The `equals` here just tries to workaround the problem of being unable to compare `std::any`, and it is expected that each type only have one comparator. It is a mistake for using multiple comparators for the same type, but unfortunately there is no way to enforce that.

Reply:
Added a `TORCH_INTERNAL_ASSERT`